README.md
# (MODX)EvolutionCMS.snippets.ddStash

Save data as [JSON](https://en.wikipedia.org/wiki/JSON) or [Query string](https://en.wikipedia.org/wiki/Query_string), then extend if needed and use it later without database queries.


## Requires

* PHP >= 5.6
* [(MODX)EvolutionCMS](https://github.com/evolution-cms/evolution) >= 1.1
* [(MODX)EvolutionCMS.libraries.ddTools](http://code.divandesign.biz/modx/ddtools) >= 0.49.1


## Documentation


### Installation


#### Manually


##### 1. Elements → Snippets: Create a new snippet with the following data

1. Snippet name: `ddStash`.
2. Description: `<b>1.3</b> Save data as JSON or QueryString, then extend if needed and use it later without database queries.`.
3. Category: `Core`.
4. Parse DocBlock: `no`.
5. Snippet code (php): Insert content of the `ddStash_snippet.php` file from the archive.


###### 2. Elements → Manage Files

1. Create a new folder `assets/snippets/ddStash/`.
2. Extract the archive to the folder (except `ddStash_snippet.php`).


#### Using [(MODX)EvolutionCMS.libraries.ddInstaller](https://github.com/DivanDesign/EvolutionCMS.libraries.ddInstaller)

Just run the following PHP code in your sources or [Console](https://github.com/vanchelo/MODX-Evolution-Ajax-Console):

```php
//Include (MODX)EvolutionCMS.libraries.ddInstaller
require_once(
	$modx->getConfig('base_path') .
	'assets/libs/ddInstaller/require.php'
);

//Install (MODX)EvolutionCMS.snippets.ddStash
\DDInstaller::install([
	'url' => 'https://github.com/DivanDesign/EvolutionCMS.snippets.ddStash',
	'type' => 'snippet'
]);
```

* If `ddStash` is not exist on your site, `ddInstaller` will just install it.
* If `ddStash` is already exist on your site, `ddInstaller` will check it version and update it if needed.


### Parameters description

* `save`
	* Desctription: Data to save in stash. Nested objects are supported too, see examples below.
	* Valid values:
		* `stringJsonObject` — as [JSON](https://en.wikipedia.org/wiki/JSON)
		* `stringHjsonObject` — as [HJSON](https://hjson.github.io/)
		* `stringQueryFormated` — as [Query string](https://en.wikipedia.org/wiki/Query_string)
		* It can also be set as a native PHP object or array (e. g. for calls through `$modx->runSnippet`):
			* `arrayAssociative`
			* `object`
	* Default value: —
	
* `save_extendExisting`
	* Desctription: Extend an existing object instead of overwriting it.
	* Valid values:
		* `0`
		* `1`
	* Default value: `0`
	
* `save_extendExistingWithEmpty`
	* Desctription: Overwrite fields with empty values (see examples below).  
		The following values are considered to be empty:
		* `""` — an empty string
		* `[]` — an empty array
		* `{}` — an empty object
		* `null`
	* Valid values:
		* `0`
		* `1`
	* Default value: `1`
	
* `get`
	* Desctription: Data key for getting from stash.
	* Valid values: `string`
	* Default value: —
	
* `get_tpl`
	* Desctription: Output template.
		
		Available placeholders:
		* `[+snippetResult+]` — data from stash
		
	* Valid values:
		* `stringChunkName`
		* `string` — use inline templates starting with `@CODE:`
	* Default value: `'@CODE:[+snippetResult+]'`
	
* `storage`
	* Desctription: Data storage.
	* Valid values:
		* `'post'` — `$_POST`
		* `'session'` — `$_SESSION`
	* Default value: `'post'`


### Examples


#### Save some data

```
[[ddStash?
	&save=`{
		"userData": {
			"firstName": "John",
			"lastName": "[[ddGetDocumentField?
				&docId=`1`
				&docField=`lastName`
			]]",
			"children": [
				{
					"firstName": "Alice"
				},
				{
					"firstName": "Robert"
				}
			]
		},
		"someData": "someValue"
	}`
]]
```

	
#### Get saved data


##### You can get field value in any depth

```
[[ddStash? &get=`someData`]]
```

Returns `someValue`.

```
[[ddStash? &get=`userData.firstName`]]
```

Returns `John`.

```
[[ddStash? &get=`userData.children.0.firstName`]]
```

Returns `Alice`.


##### Also you can get objects in JSON

If field value is object or array, it will be returned in JSON format.


###### Get first John child

```
[[ddStash? &get=`userData.children.0`]]
```

Returns:

```json
{
	"firstName": "Alice"
}
```


###### Get all John children:

```
[[ddStash? &get=`userData.children`]]
```

Returns:

```json
[
	{
		"firstName": "Alice"
	},
	{
		"firstName": "Robert"
	}
]
```


###### Get all data about John

```
[[ddStash? &get=`userData`]]
```

Returns:

```json
{
	"firstName": "John",
	"lastName": "Doe",
	"children": [
		{
			"firstName": "Alice"
		},
		{
			"firstName": "Robert"
		}
	]
}
```


#### Save: Extend an existing object instead of overwriting it (``&save_extendExisting=`1` ``)

First you save some object:

```
[[ddStash?
	&save=`{
		"userData": {
			"firstName": "Chuck",
			"lastName": "Doe",
			"children": [
				{
					"firstName": "Alice"
				},
				{
					"firstName": "Robert"
				}
			]
		}
	}`
]]
```

Then if you just save object with the same key (`userData`):

```
[[ddStash?
	&save=`{
		"userData": {
			"middleName": "Ray",
			"lastName": "Norris"
		}
	}`
]]
```

The snippet will overwrite previous saved object completely:

```
[[ddStash? &get=`userData`]]
```

Returns:

```json
{
	"middleName": "Ray",
	"lastName": "Norris"
}
```

So, if you want to extend the first object just use the `save_extendExisting` parameter:

```
[[ddStash?
	&save=`{
		"userData": {
			"middleName": "Ray",
			"lastName": "Norris"
		}
	}`
	&save_extendExisting=`1`
]]
```

In this case the snippet will recursive extend the first object with the data from the second:

```
[[ddStash? &get=`userData`]]
```

Returns:

```json
{
	"firstName": "Chuck",
	"middleName": "Ray",
	"lastName": "Norris",
	"children": [
		{
			"firstName": "Alice"
		},
		{
			"firstName": "Robert"
		}
	]
}
```


#### Save: Extend without overwriting fields with empty values (``&save_extendExistingWithEmpty=`0` ``)

By default, empty field values (e. g. `''`) are treated as other values and will replace non-empty ones.

```
[[ddStash?
	&save=`{
		"userData": {
			"firstName": "John",
			"lastName": "Tesla",
			"discipline": "Electrical engineering"
		}
	}`
]]
[[ddStash?
	&save=`{
		"userData": {
			"firstName": "Nikola",
			"lastName": ""
		}
	}`
	&save_extendExisting=`1`
]]
```

Returns:

```json
{
	"firstName": "Nikola",
	"lastName": "",
	"discipline": "Electrical engineering"
}
```

Empty `lastName` from the second object replaced non-empty `lastName` from the first.

If you want to ignore empty values, just use `save_extendExistingWithEmpty` == `0`:

```php
[[ddStash?
	&save=`{
		"userData": {
			"firstName": "John",
			"lastName": "Tesla",
			"discipline": "Electrical engineering"
		}
	}`
]]
[[ddStash?
	&save=`{
		"userData": {
			"firstName": "Nikola",
			"lastName": ""
		}
	}`
	&save_extendExisting=`1`
	&save_extendExistingWithEmpty=`0`
]]
```

Returns:

```json
{
	"firstName": "Nikola",
	"lastName": "Tesla",
	"discipline": "Electrical engineering"
}
```


#### Run the snippet through `\DDTools\Snippet::runSnippet` without DB and eval

```php
//Include (MODX)EvolutionCMS.libraries.ddTools
require_once(
	$modx->getConfig('base_path') .
	'assets/libs/ddTools/modx.ddtools.class.php'
);

//Run (MODX)EvolutionCMS.snippets.ddStash
\DDTools\Snippet::runSnippet([
	'name' => 'ddStash',
	'params' => [
		'get' => 'userData.firstName'
	]
]);
```


## Links

* [Home page](https://code.divandesign.biz/modx/ddstash)
* [Telegram chat](https://t.me/dd_code)
* [Packagist](https://packagist.org/packages/dd/evolutioncms-snippets-ddstash)


<link rel="stylesheet" type="text/css" href="https://DivanDesign.ru/assets/files/ddMarkdown.css" />

CHANGELOG.md
# (MODX)EvolutionCMS.snippets.ddStash changelog


## Version 1.3 (2021-04-28)
* \* Attention! PHP >= 5.6 is required.
* \* Attention! (MODX)EvolutionCMS.libraries.ddTools >= 0.49.1 is required.
* \+ Parameters → `save`: Can also be set as [HJSON](https://hjson.github.io/) or as a native PHP object or array (e. g. for calls through `$modx->runSnippet`).
* \+ You can just call `\DDTools\Snippet::runSnippet` to run the snippet without DB and eval (see README → Examples).
* \+ `\ddStash\Snippet`: The new class. All snippet code was moved here.
* \* README:
	* \* Documentation:
		* \+ Installation → Using (MODX)EvolutionCMS.libraries.ddInstaller.
		* \* Examples: Fixed wrong JSON format.
	* \+ Links.
* \* Composer.json:
	* \+ `homepage`.
	* \+ `support`
	* \+ `authors`.
	* \* `require` → `dd/evolutioncms-libraries-ddtools`: Fixed outdated name.


## Version 1.2.1 (2020-05-08)
* \* When you try to get non existent elements, an empty string will be returned.


## Version 1.2 (2020-05-03)
* \+ Added the ability to prevent overwriting fields with empty values (the `save_extendExistingWithEmpty` parameter, see README).
* \+ README_ru.
* \+ CHANGELOG_ru.
* \* README: Small improvements.


## Version 1.1 (2020-04-29)
* \+ Added the ability to return objects and arrays in JSON format (see README.md).
* \+ Added the ability to extend an existing object instead of overwriting it (see `save_extendExisting`).
* \+ README, CHANGELOG: Style improvements.
* \* REAMDE → Parameters description → `save`: Small improvements.
* \* Composer.json:
	* \* `version`: Fixed format.
	* \+ `keywords` → `stash`.
	* \+ `require`.


## Version 1.0 (2019-10-31)
* \+ The first release.


<link rel="stylesheet" type="text/css" href="https://DivanDesign.ru/assets/files/ddMarkdown.css" />
<style>ul{list-style:none;}</style>

src/Snippet.php
<?php
namespace ddStash;

class Snippet extends \DDTools\Snippet {
	protected
		$version = '1.3.0',
		
		$params = [
			//Defaults
			'save' => [],
			'save_extendExisting' => false,
			'save_extendExistingWithEmpty' => true,
			'get' => '',
			'get_tpl' => null,
			'storage' => 'post'
		],
		
		$paramsTypes = [
			'save' => 'objectArray',
			'save_extendExisting' => 'boolean',
			'save_extendExistingWithEmpty' => 'boolean'
		]
	;
	
	private
		$storage = []
	;
	
	/**
	 * __construct
	 * @version 1.0 (2021-04-28)
	 * 
	 * @param $params {stdClass|arrayAssociative|stringJsonObject|stringHjsonObject|stringQueryFormatted}
	 */
	public function __construct($params = []){
		//Call base method
		parent::__construct($params);
		
		//Prepare storage
		switch ($this->params->storage){
			case 'session':
				$this->storage = &$_SESSION;
			break;
			
			case 'post':
			default:
				$this->storage = &$_POST;
			break;
		}
	}
	
	/**
	 * run
	 * @version 1.0.2 (2021-04-28)
	 * 
	 * @return {string}
	 */
	public function run(){
		//The snippet must return an empty string even if result is absent
		$result = '';
		
		//Save to stash
		if (!empty($this->params->save)){
			$this->run_save();
		}
		
		//Get from stash
		if (!empty($this->params->get)){
			$result = $this->run_get();
		}
		
		return $result;
	}
	
	/**
	 * run_save
	 * @version 1.0.1 (2021-04-28)
	 * 
	 * @return {void}
	 */
	private function run_save(){
		foreach (
			$this->params->save as
			$dataName =>
			$dataValue
		){
			$dataName =
				'ddStash.' .
				$dataName
			;
			
			//If need to extend existing
			if (
				$this->params->save_extendExisting &&
				isset($this->storage[$dataName])
			){
				$this->storage[$dataName] = \DDTools\ObjectTools::extend([
					'objects' => [
						$this->storage[$dataName],
						$dataValue
					],
					'overwriteWithEmpty' => $this->params->save_extendExistingWithEmpty
				]);
			}else{
				$this->storage[$dataName] = $dataValue;
			}
		}
	}
	
	/**
	 * run_get
	 * @version 1.0.1 (2021-04-28)
	 * 
	 * @return {string}
	 */
	private function run_get(){
		$result = '';
		
		//Unfolding support (e. g. `parentKey.someKey.0`)
		$keys =	explode(
			'.',
			$this->params->get
		);
		
		//Correct parent key
		$keys[0] =
			'ddStash.' .
			$keys[0]
		;
		
		//If parent exists
		if (isset($this->storage[$keys[0]])){
			$result = $this->storage;
			
			//Find needed value
			foreach (
				$keys as
				$key
			){
				//If need to see deeper
				if (is_array($result)){
					//If element exists
					if (isset($result[$key])){
						//Save it
						$result = $result[$key];
					}else{
						//Return empty string for non-existing elements
						$result = '';
						
						break;
					}
				}else{
					break;
				}
			}
		}
		
		if (
			is_object($result) ||
			is_array($result)
		){
			$result = \DDTools\ObjectTools::convertType([
				'object' => $result,
				'type' => 'stringJsonAuto'
			]);
		}
		
		if (
			//If template is used
			!empty($this->params->get_tpl) &&
			//And result is not empty
			!empty($result)
		){
			$result = \ddTools::parseText([
				'text' => \ddTools::$modx->getTpl($this->params->get_tpl),
				'data' => [
					'snippetResult' => $result
				]
			]);
		}
		
		return $result;
	}
}